include/pybind11/detail/eigen_tensor.h

```diff
@@ -0,0 +1,383 @@
+#if defined(_MSC_VER)
+#    pragma warning(push)
+#    pragma warning(disable : 4554) // Tensor.h warning
+//       C5054: operator '&': deprecated between enumerations of different types
+#elif defined(__MINGW32__)
+#    pragma GCC diagnostic push
+#    pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#endif
+
+#include <unsupported/Eigen/CXX11/Tensor>
+
+#if defined(_MSC_VER)
+#    pragma warning(pop)
+#elif defined(__MINGW32__)
+#    pragma GCC diagnostic pop
+#endif
+
+PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+
+PYBIND11_NAMESPACE_BEGIN(detail)
+
+template <typename T>
+constexpr int compute_array_flag_from_tensor() {
+    static_assert(((int) T::Layout == (int) Eigen::RowMajor)
+                      || ((int) T::Layout == (int) Eigen::ColMajor),
+                  "Layout must be row or column major");
+    return ((int) T::Layout == (int) Eigen::RowMajor) ? array::c_style : array::f_style;
+}
+
+template <typename T>
+struct eigen_tensor_helper {};
+
+template <typename Scalar_, int NumIndices_, int Options_, typename IndexType>
+struct eigen_tensor_helper<Eigen::Tensor<Scalar_, NumIndices_, Options_, IndexType>> {
+    using T = Eigen::Tensor<Scalar_, NumIndices_, Options_, IndexType>;
+    using ValidType = void;
+
+    static std::array<typename T::Index, T::NumIndices> get_shape(const T &f) {
+        return f.dimensions();
+    }
+
+    static constexpr bool
+    is_correct_shape(const std::array<typename T::Index, T::NumIndices> & /*shape*/) {
+        return true;
+    }
+
+    template <typename T>
+    struct helper {};
+
+    template <size_t... Is>
+    struct helper<index_sequence<Is...>> {
+        static constexpr auto value = concat(const_name(((void) Is, "?"))...);
+    };
+
+    static constexpr auto dimensions_descriptor
+        = helper<decltype(make_index_sequence<T::NumIndices>())>::value;
+};
+
+template <typename Scalar_, typename std::ptrdiff_t... Indices, int Options_, typename IndexType>
+struct eigen_tensor_helper<
+    Eigen::TensorFixedSize<Scalar_, Eigen::Sizes<Indices...>, Options_, IndexType>> {
+    using T = Eigen::TensorFixedSize<Scalar_, Eigen::Sizes<Indices...>, Options_, IndexType>;
+    using ValidType = void;
+
+    static constexpr std::array<typename T::Index, T::NumIndices> get_shape(const T & /*f*/) {
+        return get_shape();
+    }
+
+    static constexpr std::array<typename T::Index, T::NumIndices> get_shape() {
+        return {{Indices...}};
+    }
+
+    static bool is_correct_shape(const std::array<typename T::Index, T::NumIndices> &shape) {
+        return get_shape() == shape;
+    }
+
+    static constexpr auto dimensions_descriptor = concat(const_name<Indices>()...);
+};
+
+template <typename T>
+struct get_tensor_descriptor {
+    static constexpr auto value
+        = const_name("numpy.ndarray[") + npy_format_descriptor<typename T::Scalar>::name
+          + const_name("[") + eigen_tensor_helper<T>::dimensions_descriptor
+          + const_name("], flags.writeable, ")
```

Comment:
This PR should have some tests for the doc_strings too.

Reply:
Numpy doesn't have "official typing strings" as it's not really designed around python type hints. 

I'm fine removing the flags here, but at that point, shouldn't we also remove the dimensions and the numpy type?